### ISSUE TITLE
Gradient of reduce_prod does not support negative axis

### ISSUE THREAD

**Issue by pvanhaes** [tensorflow/tensorflow#10835]
The gradient of `reduce_prod` does not support negative axis unlike `reduce_prod` itself.
It is apparently caused by `gather` not supporting negative axes.
This code illustrates the problem.
```python
import tensorflow as tf

vars = tf.Variable([[1., 2.], [3., 4.]])
prod = tf.reduce_prod(vars, -1) # Negative axis here

tf.InteractiveSession()
tf.global_variables_initializer().run()
print(prod.eval()) # Works fine
print(tf.gradients(prod, vars)[0].eval()) # Crashes
```

**Comment by tatatodd** [comment]
Please resubmit and pay attention to the issue template (https://github.com/tensorflow/tensorflow/issues/new) . Please provide all the information it asks (e.g. platform, version, etc). Thank you.

**Comment by pvanhaes** [comment]
I have had this bug using Tensorflow 1.2.0 (CPU) both installed using pip3 and from source.
I am running MacOS 10.12.5.

**Comment by tatatodd** [comment]
Thanks for the bug report @pvanhaes.  I've reproduced the issue, and agree with your analysis.

I think there are two ways to fix this; we can either:
1. Ensure that `tf.reduce_*` resolves negative axis arguments into positive (in-range) arguments (see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_ops.py#L1190), or
2. Change `tf.gather` to allow negative indices

I think (2) sounds reasonable.  Assigning to @ebrevdo, who seems to have some familiarity with `tf.gather`.

**Comment by ebrevdo** [comment]
Option 2 does sound reasonable.  Not sure when I'll get to this though.
Want to try submitting a PR that converts negative values within the proper
range to valid positive values?

**Comment by tatatodd** [comment]
Cool, thanks for the comment @ebrevdo!

@pvanhaes, would you like to submit a PR?  I'm marking this as "contributions welcome", so that others who are interested may also find this issue.

**Comment by AnishShah** [comment]
Can I work on this?

**Comment by ebrevdo** [comment]
Yes sounds good.

**Comment by AnishShah** [comment]
After I added negative axis support for `tf.gather`, it throws error because `tf.invert_permutation` doesn't support negative axis. Should I do the same for `tf.invert_permutation`?

**Comment by AnishShah** [comment]
The main problem is in [_ProdGrad function](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_grad.py#L127)

This is the snippet of _ProdGrad. In the comment, I have added an example where this fails.
```
reduced = math_ops.cast(reduction_indices, dtypes.int32) # [-1]
idx = math_ops.range(0, array_ops.rank(op.inputs[0])) # if rank is 2, then [0, 1]
other, _ = array_ops.setdiff1d(idx, reduced)  # others = [0, 1]
perm = array_ops.concat([reduced, other], 0) # [-1, 0, 1]
```
`tf.setdiff1d` is getting a list of `range(0, rank(input))` and reduction index is negative. So, it returns a list which contains `range(0, rank(input))` and negative indices. This is used to `tf.transpose` and it throws error because the dimensions don't match. 

I hope I make sense.

I think, the best way to solve this would be in this function itself.

**Comment by AnishShah** [comment]
This output is from tensorflow `master` branch. Except `reduce_prod`, all `reduce_*` & its gradient op work perfectly for negative axis.

```
In [1]: import tensorflow as tf

In [2]: vars = tf.Variable([[1., 2.], [3., 4.]])

In [3]: tf.InteractiveSession()
2017-06-24 00:12:38.861801: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
2017-06-24 00:12:38.861832: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
2017-06-24 00:12:38.861845: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
2017-06-24 00:12:38.861854: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
2017-06-24 00:12:38.861864: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
Out[3]: <tensorflow.python.client.session.InteractiveSession at 0x7f4145d54510>

In [4]: tf.global_variables_initializer().run()

In [5]: print(tf.gradients(tf.reduce_max(vars, -1), vars)[0].eval())
[[ 0.  1.]
 [ 0.  1.]]

In [6]: print(tf.gradients(tf.reduce_min(vars, -1), vars)[0].eval())
[[ 1.  0.]
 [ 1.  0.]]

In [7]: print(tf.gradients(tf.reduce_mean(vars, -1), vars)[0].eval())
[[ 0.5  0.5]
 [ 0.5  0.5]]

In [8]: print(tf.gradients(tf.reduce_sum(vars, -1), vars)[0].eval())
[[ 1.  1.]
 [ 1.  1.]]

In [9]: print(tf.gradients(tf.reduce_logsumexp(vars, -1), vars)[0].eval())
[[ 0.26894143  0.7310586 ]
 [ 0.26894143  0.7310586 ]]

```
